### ISSUE TITLE
Initial training configuration; Observable/Conditional/Control Columns Behavior and Best Practices in TinyTimeMixer

### ISSUE THREAD

**Issue by tonvanin** [ibm-granite/granite-tsfm#370]
Dear TinyTimeMixer Developers,

First of all, thank you for your amazing work. I've been using your solution for almost a year now and truly appreciate the architecture, documentation, and flexibility you've provided — it's a fantastic work.

Over time, however, I've encountered several confusing behaviors and would greatly appreciate your guidance, particularly regarding the roles and impact of observable_columns, conditional_columns, and control_columns.

Setup Summary:
I train TinyTimeMixer models from scratch, using a custom configuration.
My data has ~200 columns, of which 10–20 are predicted (target_columns). All columns are related.
I do not have access to future values of any feature at inference time — therefore, (as in documentation) I expected to classify them as conditional_columns.

Here’s how I define column roles:

```
column_specifiers = {
    "timestamp_column": timestamp_column,
    "target_columns": target_columns,
    # Case 1: using all non-targets as observables:
    "observable_columns": [col for col in data.columns if col not in target_columns + [timestamp_column]],
    # Case 2  using all non-targets as conditional:
    # "conditional_columns": [col for col in data.columns if col not in target_columns + [timestamp_column]],
# Case 3  using all non-targets as control:
    # "control_columns": [col for col in data.columns if col not in target_columns + [timestamp_column]],
}
```

I then pass these to:

```
tsp = TimeSeriesPreprocessor(
    **column_specifiers,
    context_length=context_length,
    prediction_length=forecast_length,
    scaling=True,
    encode_categorical=False,
    scaler_type="standard"
)
```
Model config:

```
config_kwargs = dict(
    model_type="tinytimemixer",
    context_length=context_length,
    prediction_length=forecast_length,
    patch_length=patch_length,
    patch_stride=patch_stride,
    num_input_channels=len(data.columns) - 1,
    prediction_channel_indices=prediction_channel_indices,  # optional - tried both passing and not passing
    loss="mse",
    mode=MixingMode,
    decoder_mode=MixingModeDec,
    use_decoder=True,
    ...
)
```
My Questions & Observations
1. Behavior of Conditional vs Observable
When I switch input between observable conditional and control_columns— using the exact same dataset — I get significantly different training and inference performance. 
Results for observable and control are often much better than conditional_columns, even though this seems more correct given the lack of future data.

Could this be due to the preprocessor automatically making future values available for observables? I do not pass exogenous_channel_indices manually — does the preprocessor infer and propagate future values for observables by default?

2. Prediction Channel Indices: Should I Restrict?
When I do not set prediction_channel_indices, the model tends to:
Overfit the training data (lower train loss),
But perform worse on validation loss.
When I restrict predictions only to target_columns, train loss increases slightly, but validation results improve.

Is it recommended to explicitly set prediction_channel_indices = target_columns?
Or does the model automatically train only on target_columns when passed via preprocessor?

3. Decoder Mode — When to Enable?
My data channels are highly correlated (e.g. overlapping financial indicators).

Should I enable decoder_mode and MixingModeDec during pretraining?
Or is it better to first pretrain without decoder mixing, then fine-tune later with decoder_mode and a frozen backbone?

4. Column Roles During Pretraining
Should I:
Clearly define all column roles (observable, conditional, control) during initial training,
Or is it acceptable to treat all non-targets as target_columns for pretraining, and cleanly separate roles only during fine-tuning?

5. Loss Calculation: Which Channels?
When I do not specify prediction_channel_indices, does the model optimize the loss over:
All input channels?
Or only the target_columns defined in the preprocessor?

Despite spending significant time experimenting and reading available resources, I am still unsure about the intended behaviors and best practices. Any guidance or clarification would be highly appreciated!

Thank you for your support and this wonderful open-source project.

Sincerely,
Tony

**Comment by ssiegel95** [comment]
@vijaye12 @ajati can you assist @tonvanin? He raises a number of issues which I think could help us improve our documentation is this area.

**Comment by tonvanin** [comment]
Dear TinyTimeMixer Developers,

I’d like to add an important clarification based on further investigation into one of the issues I previously described — and I believe this detail could benefit from being explicitly mentioned in the documentation:

It appears that the TimeSeriesPreprocessor reorders the columns internally, placing all target channels first in the dataset it generates.
This affects the interpretation of prediction_channel_indices.

For example:

```
all_columns = ["a", "b", "c", "d", "e", "f"]
target_columns = ["c", "e", "f"]

column_specifiers = {
    "timestamp_column": timestamp_column,
    "target_columns": target_columns,
    "observable_columns":   [col for col in data.columns if col not in target_columns + [timestamp_column]]}

tsp = TimeSeriesPreprocessor( **column_specifiers, context_length=512, prediction_length=96, scaling=True, encode_categorical=False, scaler_type="standard")

prediction_channel_indices = sorted([all_columns.index(col) for col in target_columns])
print(prediction_channel_indices)  # Output: [2, 4, 5]
print(tsp.prediction_channel_indices)  # Output: [0, 1, 2]
print(tsp.exogenous_channel_indices)   # Output: [3, 4, 5]

config["prediction_channel_indices"] = tsp.prediction_channel_indices # [0, 1, 2] => correct
config["prediction_channel_indices"] = prediction_channel_indices  # [2, 4, 5] => WRONG!
```

So while I expected to preserve the original indices of the targets from the raw DataFrame, the preprocessor instead repositions the target columns at the beginning and shifts the rest accordingly.

This behavior makes tsp.prediction_channel_indices the correct reference when configuring the model. I initially passed prediction_channel_indices using original DataFrame indices — which turned out to be inconsistent with the internal data layout. This mismatch caused some of the non-target channels to be unintentionally included as targets, and crucially, those included channels had access to their future values during training, leading to unrealistically good results.

It would be helpful to document this clearly —  that the preprocessor reshuffles columns (targets first), and that model configuration should always use tsp.prediction_channel_indices and tsp.exogenous_channel_indices rather than computing them from the raw DataFrame.

Most of the problems in understanding in previous post arise from this issue in my setup, so,
Resolved Issues from previous post:

1. Target-index mismatch – solved;
2. Loss across all channels:
If I do not pass 
config["prediction_channel_indices"] = tsp.prediction_channel_indices
while training, the model will learn optimizing MSE across ALL channels, using future values of control/observable
(correct me if I wrong)
3. Still relevant
4. partially Solved: 
When pre-training without
decoder_mode = "mix_channel" OR mode = "mix_channel" OR self_attn=True 
, I should treat all non-timestamp columns as targets.  (?)

5. Solved:
without config["prediction_channel_indices"] = tsp.prediction_channel_indices loss is calculated across ALL channels 
(correct me if I wrong)

Open Questions:
6. pretrain - finetune flow 
6.1 Broader pre-training – Is it useful to train the backbone on a wider set of channels than we plan to predict, so it learns richer patterns? In that scenario a single univariate input decoder from pre-training would later decode all channels after the backbone is frozen.
6.2 mode="mix_channel" – Should this be enabled from the very start of pre-training, and channels in final target dataset should not change once we used it?
6.3 decoder_mode = "mix_channel" - better to use it after freeze, channels should not change once we used it on final target dataset?
6.4 self_attn=True – Should self-attention heads be included right from pre-training, or added only during fine-tuning with a frozen backbone on the final dataset is fine?
7. FCM
My setup has strongly inter-related channels, and some of them serve as triggers for the targets with a lag. So I want FCM to be used only with past data.
I Still need to guarantee that no future information leaks into training.

I have tried every combination of
fcm_context_length, fcm_prepend_past, and fcm_prepend_past_offset,
and it seems the only way the FCM activates is when we set tsp.exogenous_channel_indices. Yet in every scenario future values still appear to leak into training.

Is there a recommended way to prevent this leakage?
Or, in practice, is it sufficient to rely on decoder_mode="mix_channel" and skip the FCM altogether in that case?

Many thanks in advance for any guidance!
Tony

**Comment by wgifford** [comment]
Hi @tonvanin thank you for the feedback. Indeed, the columns get placed in the tensors fed to the model, based on the following ordering:
```
                target_columns,
                observable_columns,
                control_columns,
                conditional_columns,
                categorical_columns,
```
This is done in the `ForecastDFDataset`. The methods of the preprocessor (`prediction_channel_indices` and `exogenous_channel_indices`) are there to help minimize potential mismatch issues you encountered. We will try to document this more clearly.

For your remaining questions, we will try to get back to you next week.

**Comment by Claire-bx** [comment]
I'm curious about your approach to identifying overfitting. Would you mind sharing how you usually detect it in practice? After training, I can get the train_loss, but I'm not sure how to access the validation_loss. Is there a recommended way to track both losses over time and visualize them?

**Comment by tonvanin** [comment]
@Claire-bx You should be able to track them both in default output.
While using colab notebook it will look like: 
![Image](https://github.com/user-attachments/assets/ba1237ea-1d5a-4fad-abb2-b03651482d2c)

if not, I believe - you should check callbacks
```
      finetune_forecast_args = TrainingArguments(
          output_dir=os.path.join(model_folder_name, "output"),
          overwrite_output_dir=True,
          learning_rate=learning_rate,
          num_train_epochs=num_epochs,
          do_eval=True,
          eval_strategy="epoch",
          per_device_train_batch_size=globalbatchSize,
          per_device_eval_batch_size=globalbatchSize,
          dataloader_num_workers=dataloader_num_workers,
          ddp_find_unused_parameters=False,
          report_to="tensorboard",seed=SEED,
          save_strategy="epoch",
          logging_strategy="epoch",
          save_total_limit=100,
          logging_dir=os.path.join(model_folder_name, "logs"),
          load_best_model_at_end=True,
          metric_for_best_model="eval_loss",
          greater_is_better=False,
          torch_compile=True,fp16=F16
      )
    optimizer = AdamW(finetune_forecast_model.parameters(), lr=learning_rate)
    scheduler = OneCycleLR(optimizer, learning_rate, epochs=num_epochs, steps_per_epoch=math.ceil(len(train_dataset) / globalbatchSize))

    def log_progress(trainer):
        progress_logs = trainer.state.log_history
        with open(PROGRESS_LOG_FILENAME, 'a') as f:
            for log in progress_logs:
                f.write(f"{log}\n")

    finetune_forecast_trainer = Trainer(
        model=finetune_forecast_model,
        args=finetune_forecast_args,
        train_dataset=train_dataset,
        eval_dataset=valid_dataset,
        callbacks=[early_stopping_callback, tracking_callback],
        #compute_metrics=compute_metrics,
        optimizers=(optimizer, scheduler),
    )
    finetune_forecast_trainer.train()
```
